### PR TITLE
Login tracking added

### LINKED ISSUE #20: Login Tracker For admin
Track how many users are logged in ?

### PR DESCRIPTION
**Pull request by rkishan516**
Fixes #20 
How many users are logged in can be tracked now.

### CODE REVIEW

File: controllers/LoginController.php [hunk 1]
```diff
@@ -53,7 +53,8 @@ public function Auth()
       $options['iitpConnect_user_state'] = 'logged_in';
 
       $session->set($options);
-      $result = array('response' => 'success', 'text' => User::$username, 'message' => 'login successfull');
+      isOnline();
```
Comment (anuragteapot) [inline controllers/LoginController.php]: Its not a way to call the function.  --> `User::isOnline`

File: libraries/classes/User.php [hunk 2]
```diff
@@ -133,4 +133,59 @@ public static function isLoggedIn()
       return false;
     }
   }
+
+  public function isOnline()
+  {
+    $uid = $session['uid'];
+    if($session['token'] == self::$token)
```
Comment (anuragteapot) [inline libraries/classes/User.php]: `self::$token` ?

File: libraries/classes/User.php [hunk 3]
```diff
@@ -133,4 +133,59 @@ public static function isLoggedIn()
...
+    {
+      $db = new Factory;
+      $mysql = $db->getDBO();
+
+      if($mysql->query("SELECT uid FROM user_key WHERE uid=$uid"))
```
Comment (anuragteapot) [inline libraries/classes/User.php]: I think it always checks whether query runs successfully for not. You need to check num rows.

File: controllers/LoginController.php [hunk 4]
```diff
@@ -53,7 +53,8 @@ public function Auth()
       $options['iitpConnect_user_state'] = 'logged_in';
 
       $session->set($options);
-      $result = array('response' => 'success', 'text' => User::$username, 'message' => 'login successfull');
+      User::isOnline();
```
Comment (anuragteapot) [inline controllers/LoginController.php]: Handel error these of return value is false for true.

Comment (anuragteapot) [reply controllers/LoginController.php]: What if user get logged in but due to some issue `isOnline` not able to log in. So this is not good.
So send userid as a param to `isOnline` function.

Comment (anuragteapot) [reply controllers/LoginController.php]: Then finally store in session.

File: libraries/classes/User.php [hunk 5]
```diff
@@ -133,4 +133,58 @@ public static function isLoggedIn()
...
+  public function isOnline($uid)
+  {
+    $db = new Factory;
+    $mysql = $db->getDBO();
+
+    $res = $mysql->query("SELECT uid FROM user_key WHERE uid=$uid");
+
+    if($res->num_rows)
+    {
+      $sql = "UPDATE user_keys SET uid=$uid,token='$token',isLoggedin=1,time=date('Y-m-d') WHERE uid = $uid";
```
Comment (anuragteapot) [inline libraries/classes/User.php]: From where `$token`  you get ?

File: libraries/classes/User.php [hunk 6]
```diff
@@ -133,4 +133,58 @@ public static function isLoggedIn()
...
+    {
+      throw new \Exception("Error Processing Request", $mysql->connect_error);
+      return false;
+    }
+
+    $db->disconnect();
+    return true;;
+  }
+
+  public function goOffline()
```
Comment (anuragteapot) [inline libraries/classes/User.php]: this should be a Static function

File: libraries/classes/User.php [hunk 7]
```diff
@@ -133,4 +133,58 @@ public static function isLoggedIn()
       return false;
     }
   }
+
+  public function isOnline($uid,$token)
```
Comment (anuragteapot) [inline libraries/classes/User.php]: Same here

File: libraries/classes/User.php [hunk 8]
```diff
@@ -133,4 +133,58 @@ public static function isLoggedIn()
...
+      return false;
+    }
+
+    $db->disconnect();
+    return true;;
+  }
+
+  public function goOffline()
+  {
+    $uid = $session['uid'];
```
Comment (anuragteapot) [inline libraries/classes/User.php]: First get database connection. `    $db = new Factory;
    $mysql = $db->getDBO();`

File: libraries/classes/User.php [hunk 9]
```diff
@@ -133,4 +133,58 @@ public static function isLoggedIn()
...
+
+    $sql = "SELECT token FROM user_keys where uid=$uid";
+    $rslt = $mysql->query($sql);
+
+    if($session['token'] == $rslt['token'])
+    {
+      $db = new Factory;
```
Comment (anuragteapot) [inline libraries/classes/User.php]: Remove here

**Comment by anuragteapot** [comment]
@rkishan516 And please try to add result screenshot of your PR result. I not going to test and install every time. :)

**Comment by rkishan516** [comment]
screenshot from 2018-12-21 20-13-32
Database is now working

**Comment by anuragteapot** [comment]
screenshot from 2018-12-21 20-24-21

great work @rkishan516  :)

Things, I noticed.

1. You need to add  IP column in users_key. 
2. Logout not working it's not updating the status.

And don't insert in table always,  insert when new IP is detected for that user then. If user login or logout with the same IP then just update with a new token and new time stamp and insert if new IP found.

**Comment by rkishan516** [comment]
screenshot from 2018-12-22 10-28-25
screenshot from 2018-12-22 10-28-38
Working Fine but when New Ip is detected we are updating the last ip

File: libraries/classes/User.php [hunk 10]
```diff
@@ -133,4 +133,59 @@ public static function isLoggedIn()
       return false;
     }
   }
+
+  public static function isOnline($uid,$token,$address)
+  {
+    $db = new Factory;
+    $mysql = $db->getDBO();
+
+    $res = $mysql->query("SELECT uid FROM user_keys WHERE uid=$uid");
```
Comment (anuragteapot) [inline libraries/classes/User.php]: `SELECT * FROM user_keys WHERE uid=$uid AND ip=$address`
if result is equal to 1 (num_rows) then 
UPDATE `UPDATE user_keys SET uid=$uid,token='$token',isLoggedin=1 WHERE uid = $uid AND ip=$address`
ELSE
INSERT

Its great if we store different IP where user is tried to login

File: libraries/classes/User.php [hunk 11]
```diff
@@ -133,4 +133,59 @@ public static function isLoggedIn()
...
+    $session = new Session;
+    $uid = $session->get('uid');
+
+    $sql = "SELECT token FROM user_keys where uid=$uid and isLoggedin=1";
+    $rslt = $mysql->query($sql);
+    $rslt = $rslt->fetch_assoc();
+
+    if($session->get('token') == $rslt['token'])
+    {
+      $sql = "UPDATE user_keys SET isLoggedin=0 WHERE uid = $uid and isLoggedin=1";
```
Comment (anuragteapot) [inline libraries/classes/User.php]: `UPDATE user_keys SET isLoggedin=0 WHERE uid = $uid and isLoggedin=1 AND ip=$address`

File: libraries/classes/User.php [hunk 12]
```diff
@@ -133,4 +133,59 @@ public static function isLoggedIn()
...
+    {
+      throw new \Exception("Error Processing Request", $mysql->connect_error);
+      return false;
+    }
+
+    $db->disconnect();
+    return true;
+  }
+
+  public static function goOffline()
```
Comment (anuragteapot) [inline libraries/classes/User.php:176]: `goOffilne($address)`

**Comment by anuragteapot** [comment]
Also check here user login from database. `Check current session token is valid with user_key with this ip` https://github.com/NJACKWinterOfCode/iitpConnect/blob/2ab02a6b940feb5725d995927dba494cd45d5bb5/libraries/classes/User.php#L120

**Comment by anuragteapot** [comment]
Oops wrong button

File: libraries/classes/User.php (hunk 12 above)
Comment (rkishan516) [reply libraries/classes/User.php:176]: resolved without passing

File: libraries/classes/User.php [hunk 13]
```diff
@@ -119,12 +119,25 @@ public static function checkUser($value, $return = false)
...
-    if ($status != NULL && $username != NULL && $state == 'logged_in')
+    $sql = "SELECT token FROM user_keys WHERE uid=$uid AND ip='$address'";
+    $res = $mysql->query($sql);
+    if(json_encode($res) != 'false'){
+      $res = $res->fetch_assoc();
+    }
+
+    file_put_contents('debug.txt',json_encode($res));
+
+    if ($status != NULL && $username != NULL && $state == 'logged_in' && $status = $res['token'])
```
Comment (anuragteapot) [inline libraries/classes/User.php]: And `isLoggedIn` should be 1 from database,

**Review by anuragteapot — APPROVED** [review]
Great Work :+1:

**Pull request merged**